plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextToDiagramCommands.java

```diff
@@ -235,14 +230,108 @@ protected void doExecute() {
+            org.polarsys.capella.scenario.editor.dsl.textualScenario.SequenceMessage seqMessage = 
+                (org.polarsys.capella.scenario.editor.dsl.textualScenario.SequenceMessage) messageFromXtext;
+            InstanceRole source = EmbeddedEditorInstanceHelper.getInstanceRole(seqMessage.getSource());
+            InstanceRole target = EmbeddedEditorInstanceHelper.getInstanceRole(seqMessage.getTarget());
+
+            if (source != null && target != null) {
+              SequenceMessage capellaSequenceMessage = getCorrespondingCapellaSequenceMessage(scenario, seqMessage);
+              
+              // insert message ends in interaction fragments list
+              interactionFragments.add(capellaSequenceMessage.getSendingEnd());
+              interactionFragments.add(capellaSequenceMessage.getReceivingEnd());
+              
+              Execution execution = getExecutionForSequenceMessage(scenario, capellaSequenceMessage);
+              // For withExecution messages, execution end will be added when deactivate message will be found.
+              // For simple messages, execution end will be added here.
+              if (seqMessage.getExecution() == null) {      
+                interactionFragments.add(execution.getFinish());
+              } else {
+                executionEndsToProcess.add(execution.getFinish());
+              }
+
+              sequenceMessages.add(capellaSequenceMessage);
+            }
+
+          } else {
+            // This is a ParticipantDeactivationMessage, this means that the execution on the corresponding timeline
+            // finished.
+            // We must move the interaction fragment representing the execution end on the correct position in the
+            // ownedInteractionFragments ordered list
+            ParticipantDeactivation participantDeactivationMessage = (ParticipantDeactivation) messageFromXtext;
+
+            // find the timeline (instance role) of the execution that has to end. Search by participant name
+            InstanceRole instanceRole = EmbeddedEditorInstanceHelper
+                .getInstanceRole(participantDeactivationMessage.getName());
+
+            doDeactivationSequenceMessageForReorder(interactionFragments, instanceRole, executionEndsToProcess);
+          }
+        }
+        
+        // Replace sequence message list and interaction fragments list in the real scenario with the newly computed lists
+        scenario.getOwnedInteractionFragments().clear();
+        scenario.getOwnedInteractionFragments().addAll(interactionFragments);
+        
+        scenario.getOwnedMessages().clear();
+        scenario.getOwnedMessages().addAll(sequenceMessages);
+      }
+      
     });
   }
   
+  private static Execution getExecutionForSequenceMessage(Scenario scenario, SequenceMessage sequenceMessage) {
+    Execution execution = null;
+    MessageEnd re = sequenceMessage.getReceivingEnd();
+    for (TimeLapse tl : scenario.getOwnedTimeLapses()) {
```

Comment:
update variable names re, tl to something more readable (recvEnd, timelapse maybe)